File: problem/Longest_Palindromic_Substring/longestPalindrome.java
class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0)
            return "";

        int start = 0, maxLen = 1;

        for (int i = 0; i < s.length(); i++) {
            int len1 = expandAroundCenter(s, i, i);
            int len2 = expandAroundCenter(s, i, i + 1);

            int currentMax = Math.max(len1, len2);
            if (currentMax > maxLen) {
                maxLen = currentMax;
                start = i - (currentMax - 1) / 2;
            }
        }

        return s.substring(start, start + maxLen);
    }

    private int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }

    public String longestPalindrome_dp(String s) {
        int n = s.length();
        if (n == 0)
            return "";

        boolean[][] dp = new boolean[n][n];
        int start = 0, maxLen = 1;

        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }

        for (int i = 0; i < n - 1; i++) {
            if (s.charAt(i) == s.charAt(i + 1)) {
                dp[i][i + 1] = true;
                start = i;
                maxLen = 2;
            }
        }

        for (int len = 3; len <= n; len++) {
            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1;
                if (s.charAt(i) == s.charAt(j) && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    start = i;
                    maxLen = len;
                }
            }
        }

        return s.substring(start, start + maxLen);
    }

    public String longestPalindrome_manacher(String s) {
        if (s == null || s.length() == 0)
            return "";

        StringBuilder sb = new StringBuilder("^#");
        for (char c : s.toCharArray()) {
            sb.append(c).append('#');
        }
        sb.append('$');

        String processed = sb.toString();
        int n = processed.length();
        int[] lengths = new int[n];
        int center = 0, right = 0;

        for (int i = 1; i < n - 1; i++) {
            int mirror = 2 * center - i;

            if (i < right) {
                lengths[i] = Math.min(right - i, lengths[mirror]);
            }

            while (processed.charAt(i + lengths[i] + 1) == processed.charAt(i - lengths[i] - 1)) {
                lengths[i]++;
            }

            if (i + lengths[i] > right) {
                center = i;
                right = i + lengths[i];
            }
        }

        int maxLen = 0, centerIndex = 0;
        for (int i = 1; i < n - 1; i++) {
            if (lengths[i] > maxLen) {
                maxLen = lengths[i];
                centerIndex = i;
            }
        }

        int start = (centerIndex - maxLen) / 2;
        return s.substring(start, start + maxLen);
    }
}

File: problem/Longest_Palindromic_Substring/longestPalindrome.c
int expandAroundCenter(char *s, int left, int right, int len);

char *longestPalindrome(char *s)
{
    if (!s || strlen(s) == 0)
    {
        char *result = (char *)malloc(1);
        result[0] = '\0';
        return result;
    }

    int len = strlen(s);
    int start = 0, maxLen = 1;

    for (int i = 0; i < len; i++)
    {
        int len1 = expandAroundCenter(s, i, i, len);
        int len2 = expandAroundCenter(s, i, i + 1, len);

        int currentMax = (len1 > len2) ? len1 : len2;
        if (currentMax > maxLen)
        {
            maxLen = currentMax;
            start = i - (currentMax - 1) / 2;
        }
    }

    char *result = (char *)malloc((maxLen + 1) * sizeof(char));
    strncpy(result, s + start, maxLen);
    result[maxLen] = '\0';

    return result;
}

int expandAroundCenter(char *s, int left, int right, int len)
{
    while (left >= 0 && right < len && s[left] == s[right])
    {
        left--;
        right++;
    }
    return right - left - 1;
}